hummingbot/strategy/perpetual_market_making/perpetual_market_making.py

```diff
@@ -0,0 +1,975 @@
+                        and o.client_order_id in self._exit_orders.keys()
+                        and ((position.amount < 0 and o.is_buy) or (position.amount > 0 and not o.is_buy)))]
+                for old_order in old_exit_orders:
+                    self.cancel_order(self._market_info, old_order.client_order_id)
+                    self.logger().info(
+                        f"Initiated cancellation of previous take profit order {old_order.client_order_id} in favour of new take profit order.")
+                exit_order_exists = [o for o in self.active_orders if o.price == price]
+                if len(exit_order_exists) == 0:
+                    if size > 0 and price > 0:
+                        if position.amount < 0:
+                            buys.append(PriceSize(price, size))
+                        else:
+                            sells.append(PriceSize(price, size))
+        return Proposal(buys, sells)
+
+    def _should_renew_stop_loss(self, stop_loss_order: LimitOrder) -> bool:
+        stop_loss_creation_timestamp = self._exit_orders.get(stop_loss_order.client_order_id)
+        time_since_stop_loss = self.current_timestamp - stop_loss_creation_timestamp
+        return time_since_stop_loss >= self._time_between_stop_loss_orders
+
+    def stop_loss_proposal(self, mode: PositionMode, active_positions: List[Position]) -> Proposal:
+        market: ExchangeBase = self._market_info.market
+        top_ask = market.get_price(self.trading_pair, False)
+        top_bid = market.get_price(self.trading_pair, True)
+        buys = []
+        sells = []
+
+        for position in active_positions:
+            # check if stop loss order needs to be placed
+            stop_loss_price = position.entry_price * (Decimal("1") + self._stop_loss_spread) if position.amount < 0 \
+                else position.entry_price * (Decimal("1") - self._stop_loss_spread)
+            existent_stop_loss_orders = [order for order in self.active_orders
+                                         if order.client_order_id in self._exit_orders.keys()
+                                         and ((position.amount > 0 and not order.is_buy)
+                                              or (position.amount < 0 and order.is_buy))]
+            if (not existent_stop_loss_orders
+                    or (self._should_renew_stop_loss(existent_stop_loss_orders[0]))):
+                previous_stop_loss_price = None
+                for order in existent_stop_loss_orders:
+                    previous_stop_loss_price = order.price
+                    self.cancel_order(self._market_info, order.client_order_id)
+                new_price = previous_stop_loss_price or stop_loss_price
+                if (top_ask <= stop_loss_price and position.amount > 0):
+                    price = market.quantize_order_price(
+                        self.trading_pair,
+                        new_price * (Decimal(1) - self._stop_loss_slippage_buffer))
```

Comment:
I don't believe the `new_price` to be calculated as per documented in the issue. I am not seeing the `numOfStopLossOrdersCreated` being created or referenced anywhere.

> The price is determined as (position entry price * (1 plus stop_loss_spread)) * (1 plus slippage_buffer)ˆnumberOfStopLossOrdersCreated.

Reply:
The calculation is different but the result is the same. In the strategy since we always have the previous stop loss order price, we are taking that price as the base and we multiply it by (1 plus slippage_buffer). If you extrapolate that, the price of the stop loss to be created ends up being `position_entry_price * (1 plus stop_loss_spread) * (1 plus slippage_buffer1) * (1 plus slippage_buffer2) ... * (1 plus slippage_bufferN)` and since all slippage buffers are the same (that value remains constant) it can be simplified as `position_entry_price * (1 plus stop_loss_spread) * (1 plus slippage_buffer1) * N`